pkg/provisioner/filestorage/filestorage.go

```diff
@@ -0,0 +1,215 @@
+
+// NewFilesystemProvisioner creates a new file system provisioner that creates
+// filesystems using OCI File System Service.
+func NewFilesystemProvisioner(client client.ProvisionerClient) plugin.ProvisionerPlugin {
+	return &filesystemProvisioner{
+		client: client,
+	}
+}
+
+// getMountTargetFromID retrieves mountTarget from given mountTargetID
+func (fsp *filesystemProvisioner) getMountTargetFromID(ctx context.Context, mountTargetID string) (*filestorage.MountTarget, error) {
+	resp, err := fsp.client.FileStorage().GetMountTarget(ctx, filestorage.GetMountTargetRequest{
+		MountTargetId: &mountTargetID,
+	})
+	if err != nil {
+		glog.Errorf("Failed to retrieve mount point mountTargetId=%q: %v", mountTargetID, err)
+		return nil, err
+	}
+	return &resp.MountTarget, nil
+}
+
+func (fsp *filesystemProvisioner) getOrCreateMountTarget(ctx context.Context, ad *string, subnetID string, compartmentID string) (*filestorage.MountTarget, error) {
+	// Check if there already is a mount target in the existing compartment
+	glog.Infof("Looking up existing mount targets")
+	responseListMnt, err := fsp.client.FileStorage().ListMountTargets(ctx, filestorage.ListMountTargetsRequest{
+		AvailabilityDomain: ad,
+		CompartmentId:      &compartmentID,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if len(responseListMnt.Items) != 0 {
+		glog.V(4).Infof("Found mount targets to use")
+		mntTargetSummary := responseListMnt.Items[rand.Int()%len(responseListMnt.Items)]
+		target, err := fsp.getMountTargetFromID(ctx, *mntTargetSummary.Id)
+		return target, err
+	}
+	// Mount target not created, create a new one
+	resp, err := fsp.client.FileStorage().CreateMountTarget(ctx, filestorage.CreateMountTargetRequest{
+		CreateMountTargetDetails: filestorage.CreateMountTargetDetails{
+			AvailabilityDomain: ad,
+			SubnetId:           &subnetID,
+			CompartmentId:      &compartmentID,
+			DisplayName:        common.String(fmt.Sprintf("%s%s", os.Getenv(volumePrefixEnvVarName), "mnt")),
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &resp.MountTarget, err
+}
+
+func (fsp *filesystemProvisioner) Provision(options controller.VolumeOptions, ad *identity.AvailabilityDomain) (*v1.PersistentVolume, error) {
+	ctx, cancel := context.WithTimeout(fsp.client.Context(), fsp.client.Timeout())
+	defer cancel()
+
+	fileStorageClient := fsp.client.FileStorage()
+	response, err := fileStorageClient.CreateFileSystem(ctx, filestorage.CreateFileSystemRequest{
+		CreateFileSystemDetails: filestorage.CreateFileSystemDetails{
+			AvailabilityDomain: ad.Name,
+			CompartmentId:      common.String(fsp.client.CompartmentOCID()),
+			DisplayName:        common.String(fmt.Sprintf("%s%s", os.Getenv(volumePrefixEnvVarName), options.PVC.Name)),
+		},
+	})
+	if err != nil {
+		glog.Errorf("Failed to create a file system. Options:%#v, %s", options, err)
+		return nil, err
+	}
+
+	var target *filestorage.MountTarget
+	if options.Parameters[mntTargetID] == "" {
```

Comment:
Instead of having this logic in `Provision()` I'd pass `options` into `getOrCreateMountTarget` and implement it there (and saving yourself having to pass  `options.Parameters[subnetID]`)